Homework

In this homework, we'll deploy the dino or dragon model we trained in the previous homework.

Download the model from here:

https://github.com/SVizor42/ML_Zoomcamp/releases/download/dino-dragon-model/dino_dragon_10_0.899.h5


In [1]:
!wget https://github.com/SVizor42/ML_Zoomcamp/releases/download/dino-dragon-model/dino_dragon_10_0.899.h5 

--2022-11-27 11:59:08--  https://github.com/SVizor42/ML_Zoomcamp/releases/download/dino-dragon-model/dino_dragon_10_0.899.h5
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/aa6823c5-1d5f-4ad3-a5a2-86f909f4a33b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221127%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221127T105909Z&X-Amz-Expires=300&X-Amz-Signature=6e3f4106d8bfd44253877dd5fcbfcadf3e4ecd95771ad097fbbf6e3b6e022370&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=405934815&response-content-disposition=attachment%3B%20filename%3Ddino_dragon_10_0.899.h5&response-content-type=application%2Foctet-stream [following]
--2022-11-27 11:59:09--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/aa6823c5-1d5f-

In [18]:
import tensorflow as tf
import numpy as np

## Question 1

Now convert this model from Keras to TF-Lite format.

What's the size of the **converted** model?

* 21 Mb
* -> 43 Mb
* 80 Mb
* 164 Mb

In [4]:
model = tf.keras.models.load_model('dino_dragon_10_0.899.h5')

converter = tf.lite.TFLiteConverter.from_keras_model(model)

tflite_model = converter.convert()

with open('model.tflite', 'wb') as f_out:
    f_out.write(tflite_model)

!ls -lh model.tflite

2022-11-27 12:49:03.370856: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-11-27 12:49:03.370896: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2022-11-27 12:49:03.370931: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (test-XPS-13-9350): /proc/driver/nvidia/version does not exist
2022-11-27 12:49:03.371319: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


INFO:tensorflow:Assets written to: /tmp/tmpepesfo9p/assets


INFO:tensorflow:Assets written to: /tmp/tmpepesfo9p/assets
2022-11-27 12:49:05.516842: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:362] Ignored output_format.
2022-11-27 12:49:05.516875: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:365] Ignored drop_control_dependency.
2022-11-27 12:49:05.517599: I tensorflow/cc/saved_model/reader.cc:45] Reading SavedModel from: /tmp/tmpepesfo9p
2022-11-27 12:49:05.518857: I tensorflow/cc/saved_model/reader.cc:89] Reading meta graph with tags { serve }
2022-11-27 12:49:05.518875: I tensorflow/cc/saved_model/reader.cc:130] Reading SavedModel debug info (if present) from: /tmp/tmpepesfo9p
2022-11-27 12:49:05.523264: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:357] MLIR V1 optimization pass is not enabled
2022-11-27 12:49:05.524339: I tensorflow/cc/saved_model/loader.cc:229] Restoring SavedModel bundle.
2022-11-27 12:49:05.643015: I tensorflow/cc/saved_model/loader.cc:213] Running initializatio

-rw-rw-r-- 1 test test 43M Nov 27 12:49 model.tflite


## Question 2

To be able to use this model, we need to know the index of the input and 
the index of the output. 

What's the output index for this model?

* 3
* 7
* --> 13
* 24

In [6]:
import tensorflow.lite as tflite

interpreter = tflite.Interpreter(model_path='model.tflite')
interpreter.allocate_tensors()

INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


In [8]:
input_index = interpreter.get_input_details()[0]['index']
output_index = interpreter.get_output_details()[0]['index']

In [9]:
print(input_index, output_index)

0 13


## Preparing the image

You'll need some code for downloading and resizing images. You can use 
this code:

```python
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img
```

For that, you'll need to have `pillow` installed:

```bash
pip install pillow
```

Let's download and resize this image: 

https://upload.wikimedia.org/wikipedia/commons/thumb/d/df/Smaug_par_David_Demaret.jpg/1280px-Smaug_par_David_Demaret.jpg

Based on the previous homework, what should be the target size for the image?


## Question 3

Now we need to turn the image into numpy array and pre-process it. 

> Tip: Check the previous homework. What was the pre-processing 
> we did there?

After the pre-processing, what's the value in the first pixel, the R channel?

* 0.3353411
* --> 0.5529412
* 0.7458824
* 0.9654902

In [10]:
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

In [12]:
!wget https://upload.wikimedia.org/wikipedia/commons/thumb/d/df/Smaug_par_David_Demaret.jpg/1280px-Smaug_par_David_Demaret.jpg

--2022-11-27 12:55:30--  https://upload.wikimedia.org/wikipedia/commons/thumb/d/df/Smaug_par_David_Demaret.jpg/1280px-Smaug_par_David_Demaret.jpg
Resolving upload.wikimedia.org (upload.wikimedia.org)... 91.198.174.208, 2620:0:862:ed1a::2:b
Connecting to upload.wikimedia.org (upload.wikimedia.org)|91.198.174.208|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 376127 (367K) [image/jpeg]
Saving to: ‘1280px-Smaug_par_David_Demaret.jpg’

1280px-Smaug_par_Da 100%[===================>] 367,31K  2,28MB/s    in 0,2s    

2022-11-27 12:55:30 (2,28 MB/s) - ‘1280px-Smaug_par_David_Demaret.jpg’ saved [376127/376127]



In [14]:
img = download_image('https://upload.wikimedia.org/wikipedia/commons/thumb/d/df/Smaug_par_David_Demaret.jpg/1280px-Smaug_par_David_Demaret.jpg')
img = prepare_image(img, target_size=(150, 150))


/tmp/ipykernel_4340/2487457277.py:17: DeprecationWarning: NEAREST is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.NEAREST or Dither.NONE instead.
  img = img.resize(target_size, Image.NEAREST)


In [19]:
# turn img to numpy array
img = np.array(img)
img = img/255.0

array([[[0.55294118, 0.31764706, 0.17647059],
        [0.57254902, 0.34117647, 0.17647059],
        [0.56470588, 0.3254902 , 0.14901961],
        ...,
        [0.07058824, 0.02745098, 0.01960784],
        [0.07058824, 0.03921569, 0.02745098],
        [0.10980392, 0.04705882, 0.04705882]],

       [[0.51372549, 0.28235294, 0.14901961],
        [0.55294118, 0.30980392, 0.15686275],
        [0.58039216, 0.34901961, 0.18431373],
        ...,
        [0.05882353, 0.04313725, 0.03137255],
        [0.05490196, 0.02352941, 0.01176471],
        [0.09803922, 0.03529412, 0.03921569]],

       [[0.50588235, 0.29411765, 0.15294118],
        [0.56862745, 0.34117647, 0.19607843],
        [0.55686275, 0.32156863, 0.17647059],
        ...,
        [0.0745098 , 0.02745098, 0.02745098],
        [0.09019608, 0.03529412, 0.03137255],
        [0.0745098 , 0.03529412, 0.03137255]],

       ...,

       [[0.29803922, 0.41960784, 0.36078431],
        [0.4       , 0.36078431, 0.35294118],
        [0.42745098, 0

In [20]:
img[0, 0, 0]

0.5529411764705883

## Question 4

Now let's apply this model to this image. What's the output of the model?

* 0.17049132
* 0.39009996
* 0.60146114
* --> 0.82448614


In [21]:
model.predict(img[np.newaxis, ...])

1/1 [==============================] - 0s 255ms/step


array([[0.82448477]], dtype=float32)

## Prepare the lambda code 

Now you need to copy all the code into a separate python file. You will 
need to use this file for the next two questions.

Tip: you can test this file locally with `ipython` or Jupyter Notebook 
by importing the file and invoking the function from this file.  


## Docker 

For the next two questions, we'll use a Docker image that we already 
prepared. This is the Dockerfile that we used for creating the image:

```docker
FROM public.ecr.aws/lambda/python:3.9
COPY dino-vs-dragon-v2.tflite .
```

And pushed it to [`svizor42/zoomcamp-dino-dragon-lambda:v2`](https://hub.docker.com/r/svizor42/zoomcamp-dino-dragon-lambda/tags).

A few notes:

* The image already contains a model and it's not the same model
  as the one we used for questions 1-4.
* The version of Python is 3.9, so you need to use the right wheel for 
  TF-Lite. For Tensorflow 2.7.0, it's https://github.com/alexeygrigorev/tflite-aws-lambda/raw/main/tflite/tflite_runtime-2.7.0-cp39-cp39-linux_x86_64.whl





## Question 5

Download the base image `svizor42/zoomcamp-dino-dragon-lambda:v2`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 139 Mb
* 329 Mb
* --> 639 Mb
* 929 Mb

You can get this information when running `docker images` - it'll be in the "SIZE" column.


## Question 6

Now let's extend this docker image, install all the required libraries
and add the code for lambda.

You don't need to include the model in the image. It's already included. 
The name of the file with the model is `dino-vs-dragon-v2.tflite` and it's 
in the current workdir in the image (see the Dockerfile above for the 
reference).

Now run the container locally.

Score this image: https://upload.wikimedia.org/wikipedia/en/e/e9/GodzillaEncounterModel.jpg

What's the output from the model?

* 0.12
* 0.32
* 0.52
* 0.72

## Publishing it to AWS

Now you can deploy your model to AWS!

* Publish your image to ECR
* Create a lambda function in AWS, use the ECR image
* Give it more RAM and increase the timeout 
* Test it
* Expose the lambda function using API Gateway

This is optional and not graded.


## Publishing to Docker hub

This is just for reference, this is how we published our image to Docker hub:

```bash
docker build -t zoomcamp-dino-dragon-lambda .
docker tag zoomcamp-dino-dragon-lambda:latest svizor42/zoomcamp-dino-dragon-lambda:v2
docker push svizor42/zoomcamp-dino-dragon-lambda:v2
```
